### PR TITLE
Added unit tests

### PR DESCRIPTION
**Pull request by MrPointer**
This work adds a testing framework based [Catch](https://github.com/catchorg/Catch2) and [fff](https://github.com/meekrosoft/fff).
It also adds many unit-tests to the library, ensuring expected behavior.

The tests are optional and can be enabled by passing the `-DTEST=ON` and `-DUNIT_TEST=ON` CMake flags.

### CODE REVIEW

File: test/unit/src/action/MultiImplementationActionTests.cpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+// Fake declared functions to enable call statistics
+FAKE_VOID_FUNC(actionSingleArgImpl1, int)
+
+FAKE_VOID_FUNC(actionSingleArgImpl2, int)
+
+FAKE_VOID_FUNC(actionSingleArgImpl3, int)
+
+SCENARIO("Interface with single-arg action calls referenced implementation's method")
+{
+    GIVEN("Interface with no-args action")
```
Comment (RonAmihai) [inline test/unit/src/action/MultiImplementationActionTests.cpp]: Fix given to "Interface with single-arg action"

File: test/unit/src/composite/MultiImplementationCompositeTests.cpp [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+            }
+
+            THEN("Action implementation 2 is NOT called")
+            {
+                REQUIRE(compositeActionNoArgsImpl2_fake.call_count == 0);
+            }
+            THEN("Func implementation 2 is NOT called")
+            {
+                REQUIRE(compositeFuncNoArgsImpl2_fake.call_count == 0);
+            }
```
Comment (RonAmihai) [inline test/unit/src/composite/MultiImplementationCompositeTests.cpp:68]: Did you forget to check that (Func implementation 1) is NOT called?

File: test/unit/src/composite/MultiImplementationCompositeTests.cpp [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
```
Comment (RonAmihai) [inline test/unit/src/composite/MultiImplementationCompositeTests.cpp:86]: Same for here - Action implementation 2 should (or shouldn't) be checked too? 
I see this triple-check-only is common for all the next test cases so I'll not mention it again down below...

**Review by RonAmihai — COMMENTED** [review]
I need clarification about the multi-implementation composite tests... 
It seems you've skipped a single check in each test case for some reason I can't understand 

Edit: maybe you've skipped a check on the called interface because the other test suits cover those checks on single interface with both action and func? 

I just need you to state this in order to be sure this is the reason behind it...

File: test/unit/src/action/MultiImplementationActionTests.cpp (hunk 1 above)
Comment (MrPointer) [reply test/unit/src/action/MultiImplementationActionTests.cpp]: Correct

**Comment by MrPointer** [comment]
@RonAmihai No, actually I didn't skip the tests, or should I say - I've skipped them on purpose.
You're right in the edited comment - This is exactly the reason why I decided to skip the tests.
I know it can seem confusing at the beginning, maybe I should leave a comment there, but it does avoid unnecessary duplication. Besides, it's not related to the tested scenario at all.

**Pull request merged**